sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+the `EXPLICIT_SIMD` attribute macro. Subgroup size quiery within such
+functions will always return `1`.
+
+*Functor kernel*
+```cpp
+using AccTy = cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
+                                 cl::sycl::access::target::global_buffer>;
+class Functor {
+public:
+  Functor(int X, AccTy &Acc) : X(X), Acc(Acc) {}
+
+  EXPLICIT_SIMD void operator()() { Acc[0] += X; }
+
+private:
+  int X;
+  AccTy Acc;
+};
+```
+
+*Lambda kernel and function*
+```cpp
+EXPLICIT_SIMD SYCL_EXTERNAL
+void sycl_device_f(cl::sycl::global_ptr<int> ptr, sycl::intel::gpu::simd<float, 8> X) {
+  sycl::intel::gpu::flat_block_write(*ptr.get(), X);
+}
+...
+  Q.submit([&](cl::sycl::handler &Cgh) {
+    auto Acc1 = Buf1.get_access<cl::sycl::access::mode::read>(Cgh);
+    auto Acc2 = Buf2.get_access<cl::sycl::access::mode::read_write>(Cgh);
+
+    Cgh.single_task<class KernelID>([=] () EXPLICIT_SIMD {
+      sycl::intel::gpu::simd<float, 8> Val = sycl::intel::gpu::flat_block_read(Acc1.get_pointer());
+      sycl_device_f(Acc2, Val);
+    });
+  });
+}
+```
+
+## Implementation restrictions
+Current ESP implementation does not support using certain standard SYCL features
+inside explicit SIMD kernels and functions. Most of them will be eventually
+dropped. What's not supported today:
+- Mixing `EXPLICIT_SIMD` kernels with SYCL kernels in a single source
+- Local accessors. Local memory is allocated and accessed via explicit
+device-side API
+- 2D and 3D accessors
+- Constant accessors
+- `sycl::accessor::get_pointer()`. All memory accesses through an accessor are
+done via explicit APIs; e.g. `sycl::intel::gpu::block_store(acc, offset)`
+- Few others (to be documented)
+
+
+## Core Explicit SIMD programming APIs
+
+The DPC++ Explicit SIMD library defines the following classes to enhance the
+expressiveness for explicit SIMD data-parallel programming while enabling
+efficient mapping to SIMD vector operations on Intel GPU architectures.
+
+### SIMD vector class
+
+The ```sycl::intel::gpu::simd``` class is defined as a templated class for a vector
+of fundamental data elements. The fundamental data element type **__Ty** must be
+vectorizable, which can be any of the C++ character/integer/real floating-point types,
+or IEEE-754 half-precision floating-point type. The length of the vector
+is a constant expression, as determined by the template parameter **__N** in a
+given specialization.
+
+Each simd class object is mapped to a consecutive block of general register
+files (GRF).
+
+```cpp
+namespace cm {
+namespace gen {
+template <typename __Ty, int __N> class simd {
+public:
+  using value_type = simd<__Ty, __N, 0>;
+  using element_type = __Ty;
+  using vector_type = __vector_type_t<__Ty, __N>;
+  using type = simd<__Ty, __N>;
+  static constexpr int length = __N;
+
+  // Constructors.
+  constexpr simd();
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
```

Comment:
Those should be defaulted. Why are these constexpr?

Reply:
The first is defaulted. The others are implemented differently depending on it's device or host mode. It's marked constexpr to avoid dynamic initialization when arguments are all constant expressions in some tests.

Reply:
Having only `constexpr` ctors is not overly interesting. It becomes more interesting when all operations can be constant expressions as well. This allows to pre-compute certain values and be sure it doesn't cost at runtime. It's analogous to why you want constexpr `float` & `double` even though there were not usable as template arguments.
IIUC we will then also be able to use `simd` template arguments in C++20.